src/libstore/unix/build/local-derivation-goal.cc

```diff
@@ -1970,6 +1970,9 @@ void LocalDerivationGoal::runChild()
             if (rmdir("real-root") == -1)
                 throw SysError("cannot remove real-root directory");
 
+            // Make build root read-only, so `mkdir /homeless-shelter` would fail.
+            chmod_("/", 0555);
```

Comment:
Could this break `uid-range` builds (e.g. `tests/nixos/containers/systemd-nspawn.nix`)? Those do expect the root to be writable.

Reply:
Hmm, I guess it could.

I had another idea. If we actually `mkdir /homeless-shelter` and then make it read-only, and adjust the "home directory exists" error to be only triggered if the directory exists and is non-empty, I think it may solve the issue. Would you object to always having an empty `/homeless-shelter` directory in the sandbox?

Reply:
Perhaps this should then be controlled by a derivation-provided setting like an ["advanced attribute"](https://nix.dev/manual/nix/2.24/language/advanced-attributes)?